include/cetl/any.hpp

```diff
@@ -0,0 +1,345 @@
+        if (other.has_value())
+        {
+            other.handle(detail::action::Copy, this);
+        }
+    }
+
+    any(any&& other) noexcept
+    {
+        if (other.has_value())
+        {
+            other.handle(detail::action::Move, this);
+        }
+    }
+
+    template <typename ValueType,
+              typename Tp = std::decay_t<ValueType>,
+              typename    = std::enable_if_t<!std::is_same<Tp, any>::value &&
+                                          !cetl::pf17::detail::is_in_place_type<ValueType>::value>>
+    any(ValueType&& value)
+    {
+        soo_handler<Tp>::create(*this, std::forward<ValueType>(value));
+    }
+
+    template <typename ValueType, typename... Args, typename Tp = std::decay_t<ValueType>>
+    explicit any(cetl::in_place_type_t<ValueType>, Args&&... args)
+    {
+        soo_handler<Tp>::create(*this, std::forward<Args>(args)...);
+    }
+
+    // TODO: Add ctor#6 with `std::initializer_list`.
+
+    ~any()
+    {
+        reset();
+    }
+
+    any& operator=(const any& rhs)
+    {
+        any(rhs).swap(*this);
+        return *this;
+    }
+
+    any& operator=(any&& rhs) noexcept
+    {
+        any(std::move(rhs)).swap(*this);
+        return *this;
+    }
+
+    template <typename ValueType,
+              typename Tp = std::decay_t<ValueType>,
+              typename    = std::enable_if_t<!std::is_same<Tp, any>::value>>
+    any& operator=(ValueType&& value)
+    {
+        any(std::forward<ValueType>(value)).swap(*this);
+        return *this;
+    }
+
+    template <typename ValueType, typename... Args, typename Tp = std::decay_t<ValueType>>
+    Tp& emplace(Args&&... args)
+    {
+        reset();
+        return soo_handler<Tp>::create(*this, std::forward<Args>(args)...);
+    }
+
+    // TODO: Add emplace#2 with `std::initializer_list`.
+
+    void reset() noexcept
+    {
+        handle(detail::action::Destroy);
+    }
+
+    void swap(any& rhs) noexcept
+    {
+        if (this == &rhs)
+        {
+            return;
+        }
+
+        if (has_value() && rhs.has_value())
+        {
+            any tmp;
+            rhs.handle(detail::action::Move, &tmp);
+            handle(detail::action::Move, &rhs);
+            tmp.handle(detail::action::Move, this);
+        }
+        else if (has_value())
+        {
+            handle(detail::action::Move, &rhs);
+        }
+        else if (rhs.has_value())
+        {
+            rhs.handle(detail::action::Move, this);
+        }
+    }
+
+    CETL_NODISCARD bool has_value() const noexcept
+    {
+        return handler_ != nullptr;
+    }
+
+private:
+    // Type-erased handler.
+    using any_handler = void* (*) (detail::action, const any* /*self*/, any* /*other*/);
+
+    // Small Object Optimization (SOO) handler.
+    template <typename Tp>
+    struct soo_handler
+    {
+        static_assert(sizeof(Tp) <= Footprint, "Enlarge the footprint");
+
+        static void* handle(detail::action action, const any* self, any* other)
```

Comment:
The way I read it, non-const function parameters that are not mutated violate AUTOSAR A7-1-1.